File: all-submissions/0022-generate-parentheses/solution.java
class Solution {

    List<String> result = new ArrayList<>();

    public void backtrack(StringBuilder s, int leftCount, int rightCount, int n) {
        if(s.length() == n*2) {
            result.add(s.toString());
            return;
        } 
        if (leftCount < n) {
            s.append("(");
            backtrack(s, leftCount+1, rightCount, n);
            s.deleteCharAt(s.length()-1);
        }

        if(leftCount > rightCount) {
            s.append(")");
            backtrack(s, leftCount, rightCount+1, n);
            s.deleteCharAt(s.length()-1);
        }


        

    }
    public List<String> generateParenthesis(int n) {
        backtrack(new StringBuilder(), 0, 0, n);
        return result;
    }
}

File: all-submissions/0022-generate-parentheses/solution.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        res= []
        def backtrack(openCount, closeCount, arr):
            if len(arr) == 2*n:
                res.append("".join(arr))
                return
            if openCount < n:
                arr.append("(")
                backtrack(openCount+1, closeCount, arr)
                arr.pop()
            if openCount > closeCount:
                arr.append(")")
                backtrack(openCount, closeCount+1, arr)
                arr.pop()
        backtrack(0,0,[])
        return res
            
